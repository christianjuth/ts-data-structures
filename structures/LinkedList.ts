/**
 * LinkedList
 *
 * @summary doubly linked list
 * @author Christian Juth
 *
 * Created at     : 2019-11-21 13:17:06 
 * Last modified  : 2019-11-21 21:00:53
 */

 
interface Node<T> {
  value: T;
  next: Node<T> | null;
  prev: Node<T> | null;
}

class Node<T> implements Node<T>{
  constructor({ value, next, prev }: { value: T, next?: Node<T> | null, prev?: Node<T> | null }) {
    this.value = value;
    // turn undefined into null
    this.next = next || null;
    this.prev = prev || null;
  }
}



interface LinkedList<T> {
  front: Node<T> | null,
  back: Node<T> | null
}

class LinkedList<T> implements LinkedList<T>{

  constructor() {
    this.front = null;
    this.back = null;
  }
  
  shift(value: any): void {
    const node = new Node<T>({ 
      value,
      prev: this.front
    });
    if(this.front !== null) {
      this.front.next = node;
    } else {
      this.back = node;
    }
    this.front = node;
  }

  unshift(): T | undefined {
    if(!this.front) return;
    const output = this.front.value;
    this.front = this.front.prev;
    if(this.front !== null) {
      this.front.next = null;
    }
    return output;
  }

  push(value: T): void {
    const node = new Node<T>({ 
      value,
      next: this.back
    });
    if(this.back !== null) {
      this.back.prev = node;
    } else {
      this.front = node;
    }
    this.back = node;
  }

  pop(): T | undefined {
    if(!this.back) return;
    const output = this.back.value;
    this.back = this.back.next;
    if(this.back !== null) {
      this.back.prev = null;
    }
    return output;
  }

  remove(value: T): boolean {
    for(let crnt = this.front; crnt; crnt = crnt.prev) {
      if(crnt.value === value) {
        // front
        if(crnt.next === null) {
          this.front = crnt.prev;
          if(this.front) this.front.next = null;
        } 
        // back
        else if(crnt.prev === null) {
          this.back = crnt.next;
          if(this.back) this.back.prev = null;
        } 
        // middle
        else {
          let prev = crnt.prev,
            next = crnt.next;
          prev.next = next;
          next.prev = prev;
        }
        return true;
      }
    }
    return false;
  }

  removeAll(value: T): boolean {
    let output = false;
    for(let crnt = this.front; crnt; crnt = crnt.prev) {
      if(crnt.value === value) {
        // front
        if(crnt.next === null) {
          this.front = crnt.prev;
          if(this.front) this.front.next = null;
        } 
        // back
        else if(crnt.prev === null) {
          this.back = crnt.next;
          if(this.back) this.back.prev = null;
        } 
        // middle
        else {
          let prev = crnt.prev,
            next = crnt.next;
          prev.next = next;
          next.prev = prev;
        }
        output = true;
      }
    }
    return output;
  }

  last(): T | null {
    if(this.back) return this.back.value;
    return null;
  }

  first(): T | null {
    if(this.front) return this.front.value;
    return null;
  }

  length(): number {
    let i = 0;
    for(let crnt = this.front; crnt; crnt = crnt.prev) {
      i++;
    }
    return i;
  }

  forEach(callback: (item: T, index: number) => void): void {
    let i = 0;
    for(let n = this.front; n; n = n.prev) {
      callback(n.value, i);
      i++;
    }
  }

  contains(value: T): boolean {
    for(let n = this.front; n; n = n.prev) {
      if(n.value === value) return true;
    }
    return false;
  }

  map(callback: (item: T, index: number) => T): void {
    let i = 0;
    for(let n = this.front; n; n = n.prev) {
      n.value = callback(n.value, i);
      i++;
    }
  }

  toString(): String {
    let output = '';
    for(let n = this.front; n; n = n.prev) {
      output += (n.value + (n.prev ? ', ' : ''));
    }
    return `(${output})`;
  }

  traverse(): void {
    console.log(this.toString());
  }
}

export default LinkedList;