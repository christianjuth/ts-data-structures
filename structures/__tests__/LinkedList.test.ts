import LinkedList from '../LinkedList';


describe("LinkedList", () => {

  test("push", () => {
    let list = new LinkedList<number>();
    list.push(1);
    list.push(2);
    list.push(3);
    expect(list.length()).toBe(3);
    expect(list.toString()).toMatchSnapshot();
  });

  test("pop", () => {
    let list = new LinkedList<number>();
    list.push(1);
    list.push(2);
    list.push(3);
    expect(list.toString()).toMatchSnapshot();
    expect(list.length()).toBe(3);
    expect(list.pop()).toBe(3);
    expect(list.toString()).toMatchSnapshot();
    expect(list.length()).toBe(2);
  });

  test("shift", () => {
    let list = new LinkedList<number>();
    list.shift(3);
    list.shift(2);
    list.shift(1);
    expect(list.length()).toBe(3);
    expect(list.toString()).toMatchSnapshot();
  });

  test("unshift", () => {
    let list = new LinkedList<number>();
    list.shift(3);
    list.shift(2);
    list.shift(1);
    expect(list.toString()).toMatchSnapshot();
    expect(list.length()).toBe(3);
    expect(list.unshift()).toBe(1);
    expect(list.toString()).toMatchSnapshot();
    expect(list.length()).toBe(2);
  });

  test("forEach", () => {
    let list = new LinkedList<String>();
    list.push('2');
    list.push('3');
    list.push('4');
    list.shift('1');
    let baseline = ['1', '2', '3', '4'];
    list.forEach((item, i) => {
      expect(item).toMatch(baseline[i]);
    });
  });

  test("map", () => {
    let list = new LinkedList<number>();
    list.push(2);
    list.shift(1);
    list.push(3);
    list.map(value => {
      return value * 2
    });
    expect(list.toString()).toMatchSnapshot();
  });

  test("toString", () => {
    let list = new LinkedList<number>();
    list.push(2);
    list.push(3);
    list.shift(1);
    expect(list.toString()).toBe(('(1, 2, 3)'));
  });

  test("first", () => {
    let list = new LinkedList<number>();
    list.push(2);
    list.push(3);
    list.shift(1);
    expect(list.first()).toBe(1);
  });

  test("last", () => {
    let list = new LinkedList<number>();
    list.shift(1);
    list.push(2);
    list.push(3);
    expect(list.last()).toBe(3);
  });

  test("remove", () => {
    let list = new LinkedList<number>();
    list.shift(1);
    list.push(2);
    list.push(2);
    list.push(3);
    list.shift(0);
    list.push(4);
    expect(list.toString()).toMatchSnapshot();
    // remove from middle
    expect(list.contains(3)).toBe(true);
    expect(list.remove(3)).toBe(true);
    expect(list.contains(3)).toBe(false);
    expect(list.toString()).toMatchSnapshot();
    // remove from back
    expect(list.contains(0)).toBe(true);
    expect(list.remove(0)).toBe(true);
    expect(list.contains(0)).toBe(false);
    expect(list.toString()).toMatchSnapshot();
    // remove from front
    expect(list.contains(4)).toBe(true);
    expect(list.remove(4)).toBe(true);
    expect(list.contains(4)).toBe(false);
    expect(list.toString()).toMatchSnapshot();
     // remove one of duplicate
     expect(list.contains(2)).toBe(true);
     expect(list.remove(2)).toBe(true);
     expect(list.contains(2)).toBe(true);
     expect(list.toString()).toMatchSnapshot();
     // remove value not in array
     expect(list.remove(10)).toBe(false);
  });

  test("remove", () => {
    let list = new LinkedList<number>();
    list.shift(2);
    list.push(2);
    list.push(2);
    expect(list.toString()).toMatchSnapshot();
    // remove value not in array
    expect(list.removeAll(3)).toBe(false);
    expect(list.length()).toBe(3);
    // remove all value
    expect(list.removeAll(2)).toBe(true);
    expect(list.length()).toBe(0);
  });

});