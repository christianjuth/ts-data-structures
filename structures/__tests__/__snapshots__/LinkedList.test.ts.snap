// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkedList map 1`] = `"(2, 4, 6)"`;

exports[`LinkedList pop 1`] = `"(1, 2, 3)"`;

exports[`LinkedList pop 2`] = `"(1, 2)"`;

exports[`LinkedList push 1`] = `"(1, 2, 3)"`;

exports[`LinkedList remove 1`] = `"(0, 1, 2, 2, 3, 4)"`;

exports[`LinkedList remove 2`] = `"(0, 1, 2, 2, 4)"`;

exports[`LinkedList remove 3`] = `"(1, 2, 2, 4)"`;

exports[`LinkedList remove 4`] = `"(1, 2, 2)"`;

exports[`LinkedList remove 5`] = `"(1, 2)"`;

exports[`LinkedList remove 6`] = `"(2, 2, 2)"`;

exports[`LinkedList shift 1`] = `"(1, 2, 3)"`;

exports[`LinkedList unshift 1`] = `"(1, 2, 3)"`;

exports[`LinkedList unshift 2`] = `"(2, 3)"`;
